Implement font-synthesis longhand properties

vitorroriz opened the pull request:
#### b483b294a91df986fc660018ab160ed799be28ab
<pre>
Implement font-synthesis longhand properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=232009">https://bugs.webkit.org/show_bug.cgi?id=232009</a>
rdar://84706153

Reviewed by Tim Nguyen.

Adding support for font-synthesis longhand properties:
- font-synthesis-weight
- font-synthesis-style
- font-synthesis-small-caps

Refactoring font-synthesis shorthand so it is implemented by
using its longhands.

* LayoutTests/fast/css/getComputedStyle/computed-style-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-font-family-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt:
* LayoutTests/fast/css/getComputedStyle/resources/property-names.js:
* LayoutTests/fast/css3-text/font-synthesis-parse-expected.txt:
* LayoutTests/fast/css3-text/font-synthesis-parse.html:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-small-caps-invalid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-small-caps-invalid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-small-caps-valid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-small-caps-valid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-style-invalid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-style-invalid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-style-valid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-style-valid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-weight-invalid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-weight-invalid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-weight-valid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-weight-valid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/parsing/marker-supported-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/parsing/marker-supported-properties-in-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/svg/css/getComputedStyle-basic-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::operator FontSynthesisLonghandValue const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSSegmentedFontFace.cpp:
(WebCore::CSSSegmentedFontFace::fontRanges):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSynthesisFromStyle):
(WebCore::fontSynthesisLonghandToCSSValue):
(WebCore::fontSynthesisWeightFromStyle):
(WebCore::fontSynthesisStyleFromStyle):
(WebCore::fontSynthesisSmallCapsFromStyle):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::isCSSWideValueIDKeyword):
(WebCore::StyleProperties::getPropertyValue const):
(WebCore::StyleProperties::fontSynthesisValue const):
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeFontSynthesis):
(WebCore::consumeFontSynthesisLonghand):
(WebCore::CSSPropertyParser::parseSingleValue):
(WebCore::CSSPropertyParser::parseShorthand):
(WebCore::consumeFontSynthesis): Deleted.
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/FontCascadeCache.h:
(WebCore::FontDescriptionKey::makeFlagsKey):
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialFontSynthesisWeight):
(WebCore::FontCascadeDescription::initialFontSynthesisStyle):
(WebCore::FontCascadeDescription::initialFontSynthesisSmallCaps):
(WebCore::FontCascadeDescription::initialFontSynthesis): Deleted.
* Source/WebCore/platform/graphics/FontDescription.cpp:
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::fontSynthesisWeight const):
(WebCore::FontDescription::fontSynthesisStyle const):
(WebCore::FontDescription::fontSynthesisSmallCaps const):
(WebCore::FontDescription::hasAutoFontSynthesisWeight const):
(WebCore::FontDescription::hasAutoFontSynthesisStyle const):
(WebCore::FontDescription::hasAutoFontSynthesisSmallCaps const):
(WebCore::FontDescription::setFontSynthesisWeight):
(WebCore::FontDescription::setFontSynthesisStyle):
(WebCore::FontDescription::setFontSynthesisSmallCaps):
(WebCore::FontDescription::operator== const):
(WebCore::FontDescription::encode const):
(WebCore::FontDescription::decode):
(WebCore::FontDescription::fontSynthesis const): Deleted.
(WebCore::FontDescription::setFontSynthesis): Deleted.
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::computeNecessarySynthesis):
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::getFontPropertiesFromPattern):
* Source/WebCore/platform/text/TextFlags.h:
(): Deleted.
* Source/WebCore/style/PropertyAllowlist.cpp:
(WebCore::Style::isValidMarkerStyleProperty):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertFontSynthesis): Deleted.
* Source/WebInspectorUI/UserInterface/External/CSSDocumentation/CSSDocumentation-overrides.json:
* Source/WebInspectorUI/UserInterface/Models/CSSKeywordCompletions.js:

Canonical link: <a href="https://commits.webkit.org/255171@main">https://commits.webkit.org/255171@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81c65edc8f153c44bda330fd2f21292032f4f97f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91524 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/675 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22161 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101244 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161287 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/682 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83860 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97596 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97182 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/427 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78226 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27382 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82351 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70420 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35652 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16039 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33428 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17132 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3590 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37243 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39914 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39158 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36249 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
I'll leave this to someone else to review, but my initial thoughts are:
- we probably don't need to introduce new fields/types for the new props, and can probably just edit the bits from the shorthand itself (or we should do the opposite and only have the fields for the longhands, and not the shorthand), but having both is redundant.
- this needs tests, can you import them or create some?

anttijk commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -635,6 +635,42 @@
+        "font-synthesis-weight": {
+            "inherited": true,
+            "codegen-properties": {
+                "custom": "All",
```
I don't see why these need custom handlers, they seem like ordinary font properties that could be handled with "font-property": true codegen.

The font-synthesis shorthand has special mapping but it should be refactored so it is implemented in terms of these properties, not the other way round.

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -4116,7 +4116,7 @@ webkit.org/b/206753 imported/w3c/web-platform-tests/css/css-backgrounds/border-i
 webkit.org/b/206753 imported/w3c/web-platform-tests/css/css-backgrounds/first-letter-space-not-selected.html [ ImageOnlyFailure ]
 
-webkit.org/b/206753 imported/w3c/web-platform-tests/css/css-backgrounds/background-attachment-fixed-inside-transform-1.html [ ImageOnlyFailure ]
+ webkit.org/b/206753 imported/w3c/web-platform-tests/css/css-backgrounds/background-attachment-fixed-inside-transform-1.html [ ImageOnlyFailure ]
```
nit: unrelated change

vitorroriz replied:
ops, thank you!

vitorroriz commented on Source/WebInspectorUI/UserInterface/Models/CSSKeywordCompletions.js:
```diff
@@ -603,6 +606,15 @@ WI.CSSKeywordCompletions._propertyKeywordMap = {
     "font-synthesis": [
         "none", "weight", "style"
```
reminder: add small-caps for completion

vitorroriz commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -116,7 +116,7 @@ struct ExpansionBehavior {
 WTF::TextStream& operator<<(WTF::TextStream&, ExpansionBehavior);
 
-enum FontSynthesisValues {
+enum class FontSynthesisValues : uint8_t {
```
reminder: remove that, no longer necessary

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -626,13 +626,60 @@
+                                "font-synthesis-weight",
+                                "font-synthesis-style",
+                                "font-synthesis-small-caps"
+                             ]
```
nit: this is strangely indented. can you follow other examples in the file?

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -690,6 +692,45 @@ String StyleProperties::fontVariantValue() const
+            return String();
+
+        auto property = propertyAt(foundPropertyIndex);
+        return property.value()->cssText();
```
It's more efficient to do comparisons on the values directly rather than the cssText.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -690,6 +692,45 @@ String StyleProperties::fontVariantValue() const
+
+    // font-synthesis: none or a CSS wide value keyword
+    if (weightValue == styleValue && styleValue == capsValue)
+        return weightValue;
```
Is the check here correct if all 3 values are `auto`?

Also, once you start using the CSSValue rather than css text you can use: `value->isCSSWideKeyword()`

vitorroriz replied:
these are the values for the short-hand, so they can only be  none | [ weight || style || small-caps ]

nt1m replied:
IIUC, if you have all 3 longhands set to auto, and `weightValue == styleValue && styleValue == capsValue` would return `true`, so it would return `font-synthesis: auto` instead of `font-synthesis: weight style small-caps`. That seems wrong to me, since `font-synthesis: auto` doesn't exist.

vitorroriz replied:
you are right, fixing it!

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -690,6 +692,45 @@ String StyleProperties::fontVariantValue() const
+    if (styleValue == CSSValueAutoName)
+        appendWithPrefixIfNeeded("style");
+    if (capsValue == CSSValueAutoName)
+        appendWithPrefixIfNeeded("small-caps");
```
...and `value->valueID() == CSSValueAuto` here

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -990,12 +990,16 @@ static RefPtr<CSSValueList> consumeFontFamilyDescriptor(CSSParserTokenRange& ran
-    if (id == CSSValueNone)
-        return consumeIdent(range);
+    CSSValueID id = m_range.peek().id();
+    if (id == CSSValueNone) {
```
```suggestion
    if (m_range.peek().id() == CSSValueNone) {
```

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1007,41 +1011,39 @@ static RefPtr<CSSValue> consumeFontSynthesis(CSSParserTokenRange& range)
+}
 
-    return list;
+static RefPtr<CSSValue> consumeFontSynthesisLongHand(CSSParserTokenRange& range)
```
nit: `consumeFontSynthesisLonghand` with a lowercase `h`

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -5170,6 +5174,13 @@ bool CSSPropertyParser::parseFontFaceDescriptor(CSSPropertyID propId)
+        parsedValue = consumeFontSynthesisLongHand(m_range);
+        break;
+    case CSSPropertyFontSynthesis:
+        return consumeFontSynthesis(false);
```
This shouldn't be added, `font-synthesis` isn't part of the `@font-face` descriptor.  See https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face

nt1m commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -61,7 +61,12 @@ class FontDescription {
+    FontSynthesisLonghandValue fontSynthesisSmallCaps() const { return static_cast<FontSynthesisLonghandValue>(m_fontSynthesisCaps); }
+    bool isFontSynthesisWeightEnabled() const { return fontSynthesisWeight() == FontSynthesisLonghandValue::Auto; }
+    bool isFontSynthesisStyleEnabled() const { return fontSynthesisStyle() == FontSynthesisLonghandValue::Auto; }
+    bool isFontSynthesisSmallCapsEnabled() const { return fontSynthesisSmallCaps() == FontSynthesisLonghandValue::Auto; }
```
`Enabled` feels like the wrong word here (Enabled typically refers to some feature setting)

maybe `hasAutoFontSynthesisWeight`?

nt1m commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -160,7 +167,10 @@ class FontDescription {
     unsigned m_textRendering : 2; // TextRenderingMode
     unsigned m_script : 7; // Used to help choose an appropriate font for generic font families.
-    unsigned m_fontSynthesis : 3; // FontSynthesis type
+    // unsigned m_fontSynthesis : 3; // FontSynthesis type
```
nit: please remove this comment

nt1m commented on Source/WebInspectorUI/UserInterface/External/CSSDocumentation/CSSDocumentation.js:
```diff
@@ -1135,10 +1135,25 @@ CSSDocumentation = {
+        "description": "Controls whether user agents are allowed to synthesize small-caps font faces when a font family lacks small-caps faces.",
+        "syntax": "none | auto",
+        "url": "https://developer.mozilla.org/docs/Web/CSS/font-synthesis"
+    },
```
@dcrousso @patrickangle Does this need to be in https://searchfox.org/wubkat/source/Source/WebInspectorUI/UserInterface/External/CSSDocumentation/CSSDocumentation-overrides.json ?

dcrousso replied:
yeah please don't edit this file as it's auto-generated

nt1m commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -160,7 +167,10 @@ class FontDescription {
+    // unsigned m_fontSynthesis : 3; // FontSynthesis type
+    unsigned m_fontSynthesisWeight : 1;
+    unsigned m_fontSynthesisStyle: 1;
+    unsigned m_fontSynthesisCaps: 1;
```
```suggestion
    unsigned m_fontSynthesisStyle : 1;
    unsigned m_fontSynthesisCaps : 1;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/External/CodeMirror/css.js:
```diff
@@ -480,7 +480,7 @@ CodeMirror.defineMode("css", function(config, parserConfig) {
     "float", "float-offset", "flow-from", "flow-into", "font", "font-feature-settings",
     "font-family", "font-kerning", "font-language-override", "font-size", "font-size-adjust",
-    "font-stretch", "font-style", "font-synthesis", "font-variant",
+    "font-stretch", "font-style", "font-synthesis", "font-synthesis-weight", "font-synthesis-style", "font-synthesis-small-caps", "font-variant",
```
we shouldn't edit this file directly as it's something we periodically update/import

i vaguely recall we have a mechanism to add these kinds of things from elsewhere in Web Inspector.  does this patch work without this change?

vitorroriz replied:
Yes, I've removed it.

cdumez commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,58 @@ String StyleProperties::fontVariantValue() const
+        return property.value();
+    };
+
+    auto appendWithPrefixIfNeeded = [&](const char* word) {
```
Should be using ASCIILiteral instead of `const char*`.

nt1m commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -61,7 +61,12 @@ class FontDescription {
+    FontSynthesisLonghandValue fontSynthesisSmallCaps() const { return static_cast<FontSynthesisLonghandValue>(m_fontSynthesisCaps); }
+    bool hasFontSynthesisWeight() const { return fontSynthesisWeight() == FontSynthesisLonghandValue::Auto; }
+    bool hasFontSynthesisStyle() const { return fontSynthesisStyle() == FontSynthesisLonghandValue::Auto; }
+    bool hasFontSynthesisSmallCaps() const { return fontSynthesisSmallCaps() == FontSynthesisLonghandValue::Auto; }
```
```suggestion
    bool hasAutoFontSynthesisWeight() const { return fontSynthesisWeight() == FontSynthesisLonghandValue::Auto; }
    bool hasAutoFontSynthesisStyle() const { return fontSynthesisStyle() == FontSynthesisLonghandValue::Auto; }
    bool hasAutoFontSynthesisSmallCaps() const { return fontSynthesisSmallCaps() == FontSynthesisLonghandValue::Auto; }
```

This is the typical naming convention we use for helpers that check for `auto`, see `hasAutoAccentColor` for instance.

cdumez commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,58 @@ String StyleProperties::fontVariantValue() const
+    auto appendWithPrefixIfNeeded = [&](const char* word) {
+        if (!result.isEmpty())
+            result.append(' ');
+        result.append(String(word, strlen(word)));
```
`result.append(word)` once you switch to ASCIILiteral. No need to unnecessarily construct a String.

cdumez commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,58 @@ String StyleProperties::fontVariantValue() const
+    }
+
+    if (valueID(weightValue) == CSSValueAuto)
+        appendWithPrefixIfNeeded("weight");
```
"weight"_s to make it an ASCIILiteral

cdumez commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,58 @@ String StyleProperties::fontVariantValue() const
+    if (valueID(weightValue) == CSSValueAuto)
+        appendWithPrefixIfNeeded("weight");
+    if (valueID(styleValue) == CSSValueAuto)
+        appendWithPrefixIfNeeded("style");
```
ditto.

cdumez commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,58 @@ String StyleProperties::fontVariantValue() const
+    if (valueID(styleValue) == CSSValueAuto)
+        appendWithPrefixIfNeeded("style");
+    if (valueID(capsValue) == CSSValueAuto)
+        appendWithPrefixIfNeeded("small-caps");
```
ditto.

cdumez commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -116,15 +116,10 @@ struct ExpansionBehavior {
-    FontSynthesisWeight = 0x1,
-    FontSynthesisStyle = 0x2,
-    FontSynthesisSmallCaps = 0x4
+enum class FontSynthesisLonghandValue: uint8_t {
```
Seems we could use `: bool`?

cdumez commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -840,6 +835,14 @@ template<> struct EnumTraits<WebCore::FontOpticalSizing> {
     >;
 };
 
+template<> struct EnumTraits<WebCore::FontSynthesisLonghandValue> {
```
Then I think this would no longer be needed.

nt1m commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -116,15 +116,10 @@ struct ExpansionBehavior {
-// FIXME: Use OptionSet.
-typedef unsigned FontSynthesis;
-const unsigned FontSynthesisWidth = 3;
+enum class FontSynthesisLonghandValue: bool {
```
```suggestion
enum class FontSynthesisLonghandValue : bool {
```

nt1m commented on Source/WebInspectorUI/UserInterface/External/CSSDocumentation/CSSDocumentation.js:
```diff
@@ -313,15 +313,18 @@ CSSDocumentation = {
     },
     "align-content": {
```
the changes in this file look unrelated, can you undo them?

vitorroriz replied:
They were generated using Source/WebInspectorUI/Scripts/update-inspector-css-documentation. Should I add just the hunk related to font-synthesis?

nt1m replied:
I don't see anything related to font-synthesis in these changes (everything seems to be in overrides.json).

Possibly we should run this script in a separate commit, but this seems like the wrong place to do so. cc @dcrousso @patrickangle

vitorroriz replied:
oh my bad, I had the impression it was updated. I'm removing it and just keeping the override.

nt1m reviewed (approved):
Looks good, thanks!

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -626,13 +626,60 @@
+                             "font-synthesis-weight",
+                             "font-synthesis-style",
+                             "font-synthesis-small-caps"
+                             ]
```
Please fix up indentation here.

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2289,19 +2289,42 @@ Ref<CSSValue> ComputedStyleExtractor::fontVariantShorthandValue()
+{
+    const auto valueLonghand = style.fontDescription().fontSynthesisSmallCaps();
+    return fontSynthesisLonghandToCSSValue(valueLonghand);
+}
```
Can these be single line functions (without the intermediary variable)

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,68 @@ String StyleProperties::fontVariantValue() const
+            if (isCSSWideValueKeyword(weightValue->cssText()) || weightValueID == CSSValueNone)
+                return weightValue->cssText();
+        }
+    }
```
This would be shorter if you just compare value IDs:

```suggestion
    if (weightValueID == styleValueID && weightValueID == capsValueID) {
        // font-synthesis: value can be none or CSS wide value keyword
        if (isCSSWideValueKeyword(weightValue->cssText()) || weightValueID == CSSValueNone)
            return weightValue->cssText();
    }
```

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,68 @@ String StyleProperties::fontVariantValue() const
+
+    auto valueID = [&](CSSValue *value) {
+        if (!value)
+            return CSSValueInvalid;
```
```suggestion
        if (!value || !is<CSSPrimitiveValue>(value))
            return CSSValueInvalid;
```
to avoid crashing if we don't have a `CSSPrimitiveValue`.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,68 @@ String StyleProperties::fontVariantValue() const
+            appendWithPrefixIfNeeded("style"_s);
+        if (valueID(capsValue) == CSSValueAuto)
+            appendWithPrefixIfNeeded("small-caps"_s);
+    }
```
You shouldn't need the `isAutoOrNone` checks, since the value would just be rejected as parse-time if any other value is used, and serialization would just be empty.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,68 @@ String StyleProperties::fontVariantValue() const
+            result.append(' ');
+        result.append(word);
+    };
+    
```
Can you move this down closer to where it's used? e.g. before `if (valueID(weightValue) == CSSValueAuto)`

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -67,6 +67,11 @@ static bool isCSSWideValueKeyword(StringView value)
+{
+    return valueID == CSSValueInitial || valueID == CSSValueInherit || valueID == CSSValueUnset || valueID == CSSValueRevert || valueID == CSSValueRevertLayer;
+}
+
```
This function already exists here: https://searchfox.org/wubkat/rev/cd711e64b46522100059a32e3025f7875e4861c0/Source/WebCore/css/parser/CSSParserIdioms.h#63-75

Can we re-use that instead?

Or if we can't, I would rather that we check `value->isCSSWideKeyword()` instead (against the CSSValue, not the ID). We shouldn't introduce yet another function that does this (there's already quite a few, isUniversalKeyword, isCSSWideKeyword, etc.)

nt1m reviewed (approved):
r=me but please address my comment.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,64 @@ String StyleProperties::fontVariantValue() const
+    auto capsValueID = valueID(capsValue);
+
+    if (weightValueID != CSSValueInvalid && weightValueID == styleValueID && weightValueID == capsValueID) {
+        // font-synthesis: value can be none or CSS wide value keyword
```
```suggestion
        // Handle `none` or CSS wide-keywords.
```
nit: WebKit comment style starts with uppercase, and ends with `.`

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,64 @@ String StyleProperties::fontVariantValue() const
+            return weightValue->cssText();
+    }
+
+    // if one of the longhands is a css wide keyword but not all of them are, this is not a valid shorthand
```
```suggestion
    // If one of the longhands is a CSS-wide keyword but not all of them are, this is not a valid shorthand.
```

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -705,6 +707,64 @@ String StyleProperties::fontVariantValue() const
+        result.append(word);
+    };
+
+    // all combinations of weight style small-caps
```
This comment is probably unnecessary.

vitorroriz committed:
Implement font-synthesis longhand properties
https://bugs.webkit.org/show_bug.cgi?id=232009
rdar://84706153

Reviewed by Tim Nguyen.

Adding support for font-synthesis longhand properties:
- font-synthesis-weight
- font-synthesis-style
- font-synthesis-small-caps

Refactoring font-synthesis shorthand so it is implemented by
using its longhands.

* LayoutTests/fast/css/getComputedStyle/computed-style-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-font-family-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt:
* LayoutTests/fast/css/getComputedStyle/resources/property-names.js:
* LayoutTests/fast/css3-text/font-synthesis-parse-expected.txt:
* LayoutTests/fast/css3-text/font-synthesis-parse.html:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-small-caps-invalid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-small-caps-invalid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-small-caps-valid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-small-caps-valid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-style-invalid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-style-invalid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-style-valid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-style-valid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-weight-invalid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-weight-invalid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-weight-valid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-synthesis-weight-valid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/parsing/marker-supported-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/parsing/marker-supported-properties-in-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-pseudo-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/svg/css/getComputedStyle-basic-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::operator FontSynthesisLonghandValue const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSSegmentedFontFace.cpp:
(WebCore::CSSSegmentedFontFace::fontRanges):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSynthesisFromStyle):
(WebCore::fontSynthesisLonghandToCSSValue):
(WebCore::fontSynthesisWeightFromStyle):
(WebCore::fontSynthesisStyleFromStyle):
(WebCore::fontSynthesisSmallCapsFromStyle):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::isCSSWideValueIDKeyword):
(WebCore::StyleProperties::getPropertyValue const):
(WebCore::StyleProperties::fontSynthesisValue const):
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeFontSynthesis):
(WebCore::consumeFontSynthesisLonghand):
(WebCore::CSSPropertyParser::parseSingleValue):
(WebCore::CSSPropertyParser::parseShorthand):
(WebCore::consumeFontSynthesis): Deleted.
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/FontCascadeCache.h:
(WebCore::FontDescriptionKey::makeFlagsKey):
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialFontSynthesisWeight):
(WebCore::FontCascadeDescription::initialFontSynthesisStyle):
(WebCore::FontCascadeDescription::initialFontSynthesisSmallCaps):
(WebCore::FontCascadeDescription::initialFontSynthesis): Deleted.
* Source/WebCore/platform/graphics/FontDescription.cpp:
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::fontSynthesisWeight const):
(WebCore::FontDescription::fontSynthesisStyle const):
(WebCore::FontDescription::fontSynthesisSmallCaps const):
(WebCore::FontDescription::hasAutoFontSynthesisWeight const):
(WebCore::FontDescription::hasAutoFontSynthesisStyle const):
(WebCore::FontDescription::hasAutoFontSynthesisSmallCaps const):
(WebCore::FontDescription::setFontSynthesisWeight):
(WebCore::FontDescription::setFontSynthesisStyle):
(WebCore::FontDescription::setFontSynthesisSmallCaps):
(WebCore::FontDescription::operator== const):
(WebCore::FontDescription::encode const):
(WebCore::FontDescription::decode):
(WebCore::FontDescription::fontSynthesis const): Deleted.
(WebCore::FontDescription::setFontSynthesis): Deleted.
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::computeNecessarySynthesis):
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::getFontPropertiesFromPattern):
* Source/WebCore/platform/text/TextFlags.h:
(): Deleted.
* Source/WebCore/style/PropertyAllowlist.cpp:
(WebCore::Style::isValidMarkerStyleProperty):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertFontSynthesis): Deleted.
* Source/WebInspectorUI/UserInterface/External/CSSDocumentation/CSSDocumentation-overrides.json:
* Source/WebInspectorUI/UserInterface/Models/CSSKeywordCompletions.js:

Canonical link: https://commits.webkit.org/255171@main
